File: scripts/guardar_clima.php
<?php
//script para guardar predicciones de clima bajados desde la weather channel

$key_ID = 'a0c04db7ae160e62';
$coord_aeropuerto = '-26.84086037,-65.10494232';
$db_params_local = array(
	'host' => "localhost",
	'user' => "root",
	'pass' => "123456",
	'name' => "clima_wu",
);
$db_params_5jelly = array(
	'host' => "mysql.5jelly.com",
	'user' => "u690801609_clima",
	'pass' => "u690801609_clima",
	'name' => "u690801609_clima",
);

// pedir predicciones a 10 dias a nivel horario
function prediccion10DiasHorario($key_ID, $coord){
	$cWU = curl_init('http://api.wunderground.com/api/'.$key_ID.'/hourly10day/q/'.$coord.'.json');
	if (!$cWU){return false;}
	curl_setopt($cWU, CURLOPT_RETURNTRANSFER, TRUE);
	$json_string = curl_exec($cWU);
	curl_close($cWU);
	if (!$json_string){return false;}
	$parsed_json = json_decode($json_string, true);
	return $parsed_json;
}

// conectar a una base de datos MySQL
function conectarMySQL($params){
	$mysqli = new mysqli($params['host'], $params['user'], $params['pass'], $params['name']);
	return $mysqli;
}

// insertar una prediccion
function insertarForecast($db, $est, $obs, $idCapture){
	$date =      "'".date("Y-m-d H:i:s")."'";

	$temp =      ($obs['temp']['metric']      && $obs['temp']['metric']>-999)      ? ("'".$obs['temp']['metric']."'")      : 'NULL';
	$wspd =      ($obs['wspd']['metric']      && $obs['wspd']['metric']>-999)      ? ("'".$obs['wspd']['metric']."'")      : 'NULL';
	$wdir =      ($obs['wdir']['degrees']     && $obs['wdir']['degrees']>-999)     ? ("'".$obs['wdir']['degrees']."'")     : 'NULL';
	$humidity =  ($obs['humidity']            && $obs['humidity']>-999)            ? ("'".$obs['humidity']."'")            : 'NULL';
	$windchill = ($obs['windchill']['metric'] && $obs['windchill']['metric']>-999) ? ("'".$obs['windchill']['metric']."'") : 'NULL';
	$heatindex = ($obs['heatindex']['metric'] && $obs['heatindex']['metric']>-999) ? ("'".$obs['heatindex']['metric']."'") : 'NULL';
	$pop =       ($obs['pop']                 && $obs['pop'] > -999)               ? ("'".$obs['pop']."'")                 : 'NULL';
	$mslp =      ($obs['mslp']['metric']      && $obs['mslp']['metric']>-999)      ? ("'".$obs['mslp']['metric']."'")      : 'NULL';
	
	$fields = "idCapture, idStation, dates, temp, wspd, wdir, humidity, windchill, heatindex, pop, mslp";
	$query = sprintf("INSERT INTO clima_wu.forecasts (%s) VALUES (%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s)", 
		$fields, $idCapture, $est, $date, $temp, $wspd, $wdir, $humidity, $windchill, $heatindex, $pop, $mslp);

	return $db->query($query);
}

// insertar una captura para predicciones
function insertarCaptura($db, $est){
	$date = "'".date("Y-m-d H:i:s")."'";
	$fields = "idStation, dates";
	$query = sprintf("INSERT INTO captures (%s) VALUES (%s,%s)", 
		$fields, $est, $date);

	if ($db->query($query)){
		$r = $db->query("SELECT MAX(idCapture) AS 'id' FROM captures");
		if ($r->num_rows > 0){
			$fila = $r->fetch_assoc();
			return $fila['id'];
		}
	}
	return false;
}

function test($db_params, $key_ID, $coord, $est=0){
	echo "Iniciando pruebas...";
	echo '<br/>';
		$db = conectarMySQL($db_params);
		echo ($db->connect_errno) ? "Fallo al contenctar a MySQL: (" . $db->connect_errno . ") " . $db->connect_error : "Conexion establecida con exito.";
	echo '<br/>';
		$parsed_json = prediccion10DiasHorario($key_ID, $coord);
		echo ($parsed_json) ? 'Pedido realizado con exito.' : 'No se pudo hacer la peticion JSON.';
	echo '<br/>';
		$idCapture = insertarCaptura($db, $est);
		echo 'Resultado de guardar captura: '.(($idCapture)?$idCapture:'falso');
	echo '<br/>';
		echo 'Resultados de guardar pronosticos: <br/>';
		foreach($parsed_json['hourly_forecast'] as $obs){
			// insertarla en la base de datos
			$res = insertarForecast($db, $est, $obs, $idCapture);
			echo "  ".(($res)?"Correcto ":"Error ")."<br/>";
			break; //para prueba es suficiente una vez
		}
	echo '<br/>';
	mysql_close($db);
	sleep( 6.5 );
	echo 'Finalizado.';
}

//bajar datos de internet y guardarlos en la base de datos
function procesarPredicciones($db_params, $key_ID, $coord, $est=1){
	$db = conectarMySQL($db_params);
	$intentos = 100;
	while ($intentos > 0){
		// pedir datos del dia
		$parsed_json = prediccion10DiasHorario($key_ID, $coord);
		sleep( 6.5 ); // LIMITACION POR USO GRATUITO
		if ($parsed_json){
			$idCapture = insertarCaptura($db, $est);
			if ($idCapture){
				$intentos = -1; //para terminar bucle
				foreach($parsed_json['hourly_forecast'] as $obs){
					$res = insertarForecast($db, $est, $obs, $idCapture);
				}
				break;
			}
			else{ // quito un intento y espero para el proximo
				$intentos--;
			}
		}
		else{ // quito un intento y espero para el proximo
			$intentos--;
		}
		echo intentos."<br/>";
	}
	mysql_close($db);
	return ($intentos<0);	
}

procesarPredicciones($db_params_local, $key_ID, $coord_aeropuerto);

File: scripts/guardar-clima.py
#!/usr/bin/env python
#script para guardar predicciones de clima bajados desde la weather channel

import urllib2
import json
import MySQLdb
import time

key_ID = 'a0c04db7ae160e62'
coord_aeropuerto = '-26.84086037,-65.10494232'

# pedir predicciones a 10 dias a nivel horario
def prediccion10DiasHorario(key_ID, coord):
	try: f = urllib2.urlopen('http://api.wunderground.com/api/'+key_ID+'/hourly10day/q/'+coord+'.json')
	except: return ''
	json_string = f.read()
	parsed_json = json.loads(json_string)
	f.close()
	return parsed_json

# conectar a una base de datos MySQL
def conectarMySQL():
	db = MySQLdb.connect(host="localhost", user="root", passwd="123456", db="clima_wu")
	return db

# insertar una prediccion
def insertarForecast(db, est, obs, idCapture):
	cur = db.cursor()

	date = "'%s-%s-%s %s:%s:00'" %(obs["FCTTIME"]["year"], obs["FCTTIME"]["mon_padded"], obs["FCTTIME"]["mday_padded"], obs["FCTTIME"]["hour_padded"], obs["FCTTIME"]["min"])
	temp = obs['temp']['metric'] if len(obs['temp']['metric'])>0  else '-999'
	wspd = obs['wspd']['metric'] if len(obs['wspd']['metric'])>0  else '-999'
	wdir = obs['wdir']['degrees'] if len(obs['wdir']['degrees'])>0  else '-999'
	humidity = obs['humidity'] if len(obs['humidity'])>0  else '-999'
	windchill = obs['windchill']['metric'] if len(obs['windchill']['metric'])>0  else '-999'
	heatindex = obs['heatindex']['metric'] if len(obs['heatindex']['metric'])>0  else '-999'
	pop = obs['pop'] if len(obs['pop'])>0  else '-999'
	mslp = obs['mslp']['metric'] if len(obs['mslp']['metric'])>0  else '-999'

	try: temp = "'%s'"%temp if float(temp)>-999  else 'NULL'
	except: temp = 'NULL'
	try: wspd = "'%s'"%wspd if float(wspd)>-999  else 'NULL'
	except: wspd = 'NULL'
	try: wdir = "'%s'"%wdir if float(wdir)>-999  else 'NULL'
	except: wdir = 'NULL'
	try: humidity = "'%s'"%humidity if float(humidity)>-999  else 'NULL'
	except: humidity = 'NULL'
	try: windchill = "'%s'"%windchill if float(windchill)>-999  else 'NULL'
	except: windchill = 'NULL'
	try: heatindex = "'%s'"%heatindex if float(heatindex)>-999  else 'NULL'
	except: heatindex = 'NULL'
	try: pop = "'%s'"%pop if float(pop)>-999  else 'NULL'
	except: pop = 'NULL'
	try: mslp = "'%s'"%mslp if float(mslp)>-999  else 'NULL'
	except: mslp = 'NULL'

	fields = "idCapture, idStation, dates, temp, wspd, wdir, humidity, windchill, heatindex, pop, mslp"
	query = "INSERT INTO clima_wu.forecasts (%s) VALUES (%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s)" % (
		fields, str(idCapture), str(est), date, temp, wspd, wdir, humidity, windchill, heatindex, pop, mslp)
	
	res = cur.execute(query)
	db.commit()
	cur.close()
	return res

# insertar una captura para predicciones
def insertarCaptura(db, est):
	cur = db.cursor()
	date = time.strftime("'%Y-%m-%d %H:%M:%S'")
	fields = "idStation, dates"
	query = "INSERT INTO clima_wu.captures (%s) VALUES (%s,%s)" % (
		fields, str(est), date)
	res = 0
	r = cur.execute(query)
	if r == 1:
		db.commit()
		r = cur.execute("SELECT MAX(idCapture) FROM clima_wu.captures")
		if r == 1:
			for row in cur.fetchall():
			    res = row[0]
	cur.close()
	return res


def procesarPredicciones():
	est = 1
	db = conectarMySQL()

	idCapture = insertarCaptura(db, est)
	if idCapture == 0:
		return False

	intentos = 100
	while intentos > 0:
		# pedir datos del dia
		parsed_json = prediccion10DiasHorario(key_ID, coord_aeropuerto)
		time.sleep( 6.5 ) # LIMITACION POR USO GRATUITO

		# si todo va bien
		if len(parsed_json) > 0:
			intentos = 0
			# por cada muestra
			for obs in parsed_json['hourly_forecast']:
				# insertarla en la base de datos
				res = insertarForecast(db, est, obs, idCapture)
				if res != 1:
					print "Hubo un error al prediccion.", obs["FCTTIME"]
					return False
		else:
			# quito un intento y espero para el proximo
			intentos = intentos - 1

		print intentos

	db.close()
	return True	

procesarPredicciones()


